### PR TITLE
MSA: improve author information gui

### PR DESCRIPTION
**Pull request by rhaschke**
This continues #187, https://github.com/v4hn/moveit/pull/1.

As discussed don't install setup_assistant.launch. 
In order to warn if the user disables generation for a file that should be (re)written, I propose the following mechanism:
* Each to-be-written file has a bitfield of information pieces it depends on (`write_on_changes`).
* During editing, MSA sets similar bitfields in a global `changes` variable (part of `config_data_`).
* If the user disables file generation, those bitfields are checked for an overlap and a warning is generated.

I implemented this mechanism for author information for now. Other bits of information should follow.

### CODE REVIEW

File: moveit_setup_assistant/src/widgets/configuration_files_widget.cpp [hunk 1]
```diff
@@ -554,13 +583,18 @@ void ConfigurationFilesWidget::changeActionDesc(int id)
 // ******************************************************************************************
 void ConfigurationFilesWidget::changeCheckedState(QListWidgetItem* item)
 {
-  QVariant index = item->data(Qt::UserRole);
+  size_t index = item->data(Qt::UserRole).toUInt();
```
Comment (davetcoleman) [inline moveit_setup_assistant/src/widgets/configuration_files_widget.cpp]: std::

File: moveit_setup_assistant/src/widgets/configuration_files_widget.cpp [hunk 2]
```diff
@@ -650,7 +691,7 @@ void ConfigurationFilesWidget::showGenFiles()
 {
 
   // Display this list in the GUI
-  for (int i = 0; i < gen_files_.size(); ++i)
+  for (size_t i = 0; i < gen_files_.size(); ++i)
```
Comment (davetcoleman) [inline moveit_setup_assistant/src/widgets/configuration_files_widget.cpp]: std::

File: moveit_setup_assistant/src/widgets/configuration_files_widget.cpp [hunk 3]
```diff
@@ -214,6 +216,7 @@ bool ConfigurationFilesWidget::loadGenFiles()
   template_path     = config_data_->appendPaths( config_data_->template_package_path_, file.file_name_);
   file.description_ = "CMake build system configuration file";
   file.gen_func_    = boost::bind(&ConfigurationFilesWidget::copyTemplate, this, template_path, _1);
+  file.write_on_changes = 0;
```
Comment (davetcoleman) [inline moveit_setup_assistant/src/widgets/configuration_files_widget.cpp]: before we merge this, I think these should be chosen now.

**Comment by davetcoleman** [comment]
I like the required-to-be-generated warning feature but I think I see a problem. If you've edited, for example, the .setup_assistant file outside of the MSA Gui and the timestamps are out of sync, the file will not be checked to be generated on the "Generate Configuration Files" widget. It will allow you to "Generate Package" without any warning that the file should be generated / checked in the GUI. Your current warning only pops up if the user actively unchecks it. I think it should also warn you if it was never checked at all

**Comment by rhaschke** [comment]
> It will allow you to "Generate Package" without any warning that the file should be generated / checked in the GUI. Your current warning only pops up if the user actively unchecks it. I think it should also warn you if it was never checked at all.

I see. What about marking externally modified files for re-creation but highlighting them in red and providing a corresponding message? Then, if the file is deselected, the user is warned again.

Regarding the other files, I'm not exactly sure, which pieces of information they depend on. I will make an initial proposal though.

**Comment by davetcoleman** [comment]
> What about marking externally modified files for re-creation but highlighting them in red and providing a corresponding message? Then, if the file is deselected, the user is warned again.

Sounds good to me

**Comment by rhaschke** [comment]
I decided to keep conflicting files deselected - explicit changes by user should still have precedence. However, they will be marked red. Also, I change the generation-dialog to re-check the file list each time the dialog is chosen - in order to accommodate user edits in between.

**Comment by rhaschke** [comment]
Preparing a rebase to resolve conflicts before merging.

**Comment by rhaschke** [comment]
Finished rebasing.

**Comment by davetcoleman** [comment]
Travis failed - looks like an integration test is failing:

```
Invalid Trajectory: start point deviates from current robot state more than 0.01
joint 'joint_1': expected: 0.188012, current: 0.199903
[ INFO] [1474331918.610170913]: Execution completed: ABORTED
terminate called after throwing an instance of 'boost::exception_detail::clone_impl<boost::exception_detail::error_info_injector<boost::lock_error> >'
  what():  boost: mutex lock failed in pthread_mutex_lock: Invalid argument
testpython_move_group ... ok
[ROSTEST]-----------------------------------------------------------------------
[moveit_ros_planning_interface.rosunit-python_move_group/test_target_setting][passed]
[moveit_ros_planning_interface.rosunit-python_move_group/test_validation][FAILURE]
False is not true
  File "/usr/lib/python2.7/unittest/case.py", line 329, in run
    testMethod()
  File "/root/ws_moveit/src/moveit/moveit_ros/planning_interface/test/python_move_group.py", line 57, in test_validation
    self.assertTrue(self.group.execute(plan3))
  File "/usr/lib/python2.7/unittest/case.py", line 422, in assertTrue
    raise self.failureException(msg)
--------------------------------------------------------------------------------
SUMMARY
 * RESULT: FAIL
 * TESTS: 2
 * ERRORS: 0
 * FAILURES: 1
```

Seems totally unrelated to this PR but indicates we have a flaky test probably because of https://github.com/ros-planning/moveit/pull/63

I'll retrigger Travis now

**Comment by davetcoleman** [comment]
Just tested your changes. I notice that the popup warning that says "Attention: Some files (marked red) are changed both, externally and in the Setup Assistant" is always shown (when there are files that have changed outside of the MSA) even when there are no files actually marked red. Perhaps that part of the popup warning is hidden unless there are files that qualify to be red? It confused me at first.

**Comment by rhaschke** [comment]
The failing Travis check is due to #232 still missing, i.e. not waiting for robot state updates before planning, trajectory validation, etc. Having #63 merged now, I re-filed this PR.

**Comment by rhaschke** [comment]
> I notice that the popup warning that says "Attention: Some files (marked red) are changed both, externally and in the Setup Assistant" is always shown - even when there are no files actually marked red.

Hm. This doesn't happen for me:

screenshot_2016-09-21_00-29-20

Could you have a look into your console to see, which files are conflicting? Maybe, you manually modified .setup_assistant? This is indeed always considered to be "conflicting". However, in this case this file turns red for me as expected:

screenshot_2016-09-21_00-33-24

**Comment by davetcoleman** [comment]
Looking closer, the problem is that I get the popup box with a red text warning *even when the file that must be written is already checked*. So if I change the Author Information then go to the Configuration Files screen, it tells me ``.setup_assistant`` must be written but its already checked and not red. I think this is somewhat confusing for users. Can we make that red text message hidden if the necessary files are already checked?
screenshot from 2016-09-21 15-25-03

File: moveit_setup_assistant/src/widgets/configuration_files_widget.cpp [hunk 4]
```diff
@@ -660,14 +715,17 @@ void ConfigurationFilesWidget::showGenFiles()
...
-      item->setCheckState(Qt::Unchecked);
-      item->setForeground( QBrush(QColor(255, 135, 0)));
+      if (file->write_on_changes & config_data_->changes)
+      {
+        have_conflicting_changes = true;
+        item->setForeground( QBrush(QColor(255, 0, 0)));
+      }
+      else
+        item->setForeground( QBrush(QColor(255, 135, 0)));
     }
```
Comment (rhaschke) [inline moveit_setup_assistant/src/widgets/configuration_files_widget.cpp]: Here and only here, the files entries are colored red or orange. If they are colored red, also the flag `have_conflicting_changes` is set which triggers the extra warning. I cannot see a reason for the behavior observed by @davetcoleman.

**Comment by rhaschke** [comment]
I cannot reproduce your behavior on my side. If the red extra warning pops up, at least one file is marked red. That's also what is stated in the code [here](https://github.com/ros-planning/moveit/pull/221/files#r80004576). Maybe, there is some build issue and you should try to clean your workspace before re-building?

**Comment by davetcoleman** [comment]
The fault was mine - I was not seeing the red because ``package.xml`` is highlighted as the first item in the list. I didn't realize my mistake because I was thinking it would show up for ``.setup_assistant`` instead

**Comment by davetcoleman** [comment]
Should this be merged into I/J also?

**Comment by v4hn** [comment]
Yes. This (should) resolve the problem I explained in https://github.com/ros-planning/moveit/issues/100#issuecomment-248659526 that also affects indigo.

**Comment by davetcoleman** [comment]
ill create PRs now

**Pull request merged**